src/permutation.jl

```diff
@@ -0,0 +1,65 @@
+export PermutationTest
+
+function ptstats(x,y,f)
+    xy = vcat(x,y)
+    rx = 1 : length(x)
+    ry = length(xy)-length(y)+1 : length(xy)
+    T = typeof(f(x))  # is there a better way?  e.g., does a funtype() exist, like eltype()??
+    (xy, rx, ry, T)
+end
+
+immutable PermutationTest <: HypothesisTests.HypothesisTest
+    observation
+    samples
+end
+
+"""
+    PermutationTest(x::Vector, y::Vector, f::Function [, n::Int])
+
+Perform a permutation test (a.k.a. randomization test) of the null hypothesis
+that the probability of `f(x)` is equal to `f(y)`.
+
+If `n` is *not* supplied, all possible permutations are sampled.  Otherwise,
+only `n` of the `factorial(length(vec(x,y)))` permutations are sampled.
+"""
+function PermutationTest{R<:Real,S<:Real}(x::AbstractVector{R}, y::AbstractVector{S}, f::Function)
+    xy, rx, ry, T = ptstats(x,y,f)
+    P = permutations(xy)
+    xsamples = Array(T,length(P))
+    ysamples = Array(T,length(P))
+    for (i,p) in enumerate(P)
+        xsamples[i] = f(p[rx])
+        ysamples[i] = f(p[ry])
+    end
```

Comment:
`@inbounds` won't make a difference since `rx` and `ry` are ranges and `p[rx]` and `p[ry]` are allocating new vectors. This would ideally be written in a way that does not allocate memory, although that's somewhat difficult because `permutations` itself allocates new vectors.